Generate serialization of Skia Font implementation

donny-dont opened the pull request:
#### aa1e67cc9e681eb2e8fea487a9bef1e542983ec6
<pre>
Generate serialization of Skia Font implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=273445">https://bugs.webkit.org/show_bug.cgi?id=273445</a>

Reviewed by Alex Christensen.

Add `WebCoreArgumentCodersSkia.serialization.in` which generates serialization
of Skia&apos;s font implementation. Follow along with the CoreText implementation
to reuse as much of their implementation as possible.

* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformDataAttributes::FontPlatformDataAttributes):
* Source/WebCore/platform/graphics/skia/FontCustomPlatformDataSkia.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::FontCustomPlatformData::serializedData const):
* Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp:
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::platformDataInit):
(WebCore::FontPlatformData::fromIPCData):
(WebCore::FontPlatformData::toIPCData const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::shouldUseLockdownFontParser const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreFont.serialization.in:
* Source/WebKit/Shared/harfbuzz/WebCoreArgumentCodersHarfBuzz.serialization.in: Copied from Source\WebKit\Shared\WebCoreFont.serialization.in.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
(IPC::ArgumentCoder&lt;WebCore::Font&gt;::encodePlatformData): Deleted.
(IPC::ArgumentCoder&lt;WebCore::Font&gt;::decodePlatformData): Deleted.
(IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::encodePlatformData): Deleted.
(IPC::ArgumentCoder&lt;WebCore::FontPlatformData::Attributes&gt;::decodePlatformData): Deleted.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.serialization.in: Copied from Source\WebKit\Shared\WebCoreFont.serialization.in.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):

Canonical link: <a href="https://commits.webkit.org/286581@main">https://commits.webkit.org/286581@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00da610979da0b8a8e55ff53d96ef95587708e09

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76391 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55426 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29297 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80926 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27667 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78508 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64568 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3719 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59902 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18018 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79458 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49804 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65601 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40234 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47199 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23090 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25989 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68327 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23422 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82364 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3767 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2465 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68114 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3920 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65570 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67428 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16815 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11389 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9490 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3714 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6523 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3737 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7167 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5495 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann reviewed (commented):
Looks good in general, thanks @donny-dont. I'd like to hear @carlosgcampos opinion on Skia API usage too.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FontCustomPlatformDataSkia.cpp:
```diff
@@ -140,4 +140,21 @@ bool FontCustomPlatformData::supportsTechnology(const FontTechnology&)
+FontCustomPlatformSerializedData FontCustomPlatformData::serializedData() const
+{
+    auto data = m_typeface->serialize();
+    std::span<const uint8_t> copy(data->bytes(), data->size());
```
Why do we need to serialize the typeface here? we already have the font data in creationData.fontFaceData.

donny-dont replied:
Good catch! Fixed!

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp:
```diff
@@ -54,6 +54,42 @@ FontPlatformData::FontPlatformData(sk_sp<SkTypeface>&& typeface, float size, boo
+
+    auto serialized = m_font.getTypeface()->serialize();
+    std::span<const uint8_t> copy(serialized->bytes(), serialized->size());
+    return { WTFMove(copy), ""_s };
```
So, we serialize the font, but it's encoded as if it was FontCustomPlatformData? I think we could add encoder/decode for sk_sp<SkTypeface> similar to the one we added to color spaces and here use a variant, to either use the data for FontCustomPlatformData or the font for FontPlatformData

donny-dont replied:
Researching this idea. Not sure if what I was doing for serialization was ok here.

donny-dont replied:
@achristensen07 noticed that this is going to leak because a `std::span` is just a pointer and size which has no ownership.

Addressing in https://github.com/WebKit/WebKit/pull/28301 and then will follow up with changes here that resemble what you were talking about.

cdumez replied:
@donny-dont I don't see any leak here? We construct a Vector from the span. The Vector copies the span data and keeps its memory alive. I don't think we need to free the value returned by serialize() either since it is returning a smart pointer.

mcatanzaro replied:
We should probably never name a `std::span` "copy" since that's not what it does?

cdumez replied:
> We should probably never name a `std::span` "copy" since that's not what it does?

That is very true :)

donny-dont replied:
So this was changed to be a variant as requested and shouldnt leak at all

cdumez commented on Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp:
```diff
@@ -54,6 +54,46 @@ FontPlatformData::FontPlatformData(sk_sp<SkTypeface>&& typeface, float size, boo
+
+FontPlatformData::IPCData FontPlatformData::toIPCData() const
+{
+    const auto& data = creationData();
```
Please use:
```
if (auto* data = creationData())
    return FontPlatformSerializedCreationData { { data->fontFaceData->span() }, data->itemInCollection };
```

Hiding a raw pointer being a reference is too sneaky. Also, reducing the scope of variables is always good, especially when using raw pointers.

donny-dont replied:
Good catch. That was just copied from the CoreText implementation.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp:
```diff
@@ -54,6 +54,45 @@ FontPlatformData::FontPlatformData(sk_sp<SkTypeface>&& typeface, float size, boo
+    m_font.setHinting(FontRenderOptions::singleton().hinting());
+
+    m_hbFont = SkiaHarfBuzzFont::getOrCreate(*m_font.getTypeface());
+}
```
What happens with the features? I think we could just use the existing constructor instead of duplicating the code here.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp:
```diff
@@ -54,6 +54,45 @@ FontPlatformData::FontPlatformData(sk_sp<SkTypeface>&& typeface, float size, boo
+        },
+        [&] (FontPlatformSerializedCreationData& d) -> std::optional<FontPlatformData> {
+            auto fontFaceData = SharedBuffer::create(WTFMove(d.fontFaceData));
+            if (RefPtr fontCustomPlatformData = FontCustomPlatformData::create(fontFaceData, d.itemInCollection))
+                return FontPlatformData(size, WTFMove(orientation), WTFMove(widthVariant), WTFMove(textRenderingMode), syntheticBold, syntheticOblique, WTFMove(fontCustomPlatformData));
```
I guess we can use the existing constructor here, getting the typeface from the custom data and also passing the features that should be serialized too, I guess.

achristensen07 commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -287,9 +308,12 @@ class FontPlatformData {
     using IPCData = std::variant<FontPlatformSerializedData, FontPlatformSerializedCreationData>;
     WEBCORE_EXPORT FontPlatformData(float size, FontOrientation&&, FontWidthVariant&&, TextRenderingMode&&, bool syntheticBold, bool syntheticOblique, RetainPtr<CTFontRef>&&, RefPtr<FontCustomPlatformData>&&);
+#elif USE(SKIA)
+    using IPCData = std::variant<FontPlatformSerializedData, FontPlatformSerializedCreationData>;
```
This can probably be shared.

achristensen07 commented on Source/WebKit/Shared/harfbuzz/WebCoreArgumentCodersHarfBuzz.serialization.in:
```diff
@@ -0,0 +1,35 @@
+[CustomHeader] struct hb_feature_t {
+     hb_tag_t tag;
+     uint32_t value;
+     unsigned int start;
```
This should probably be just "unsigned"

achristensen07 commented on Source/WebKit/Shared/WebCoreFont.serialization.in:
```diff
@@ -20,20 +20,43 @@
+using WebCore::FontPlatformData::IPCData = std::variant<WebCore::FontPlatformSerializedData, WebCore::FontPlatformSerializedCreationData>;
+#endif
+#if USE(SKIA)
+using WebCore::FontPlatformData::IPCData = std::variant<WebCore::FontPlatformSerializedData, WebCore::FontPlatformSerializedCreationData>;
```
This is the same as CoreText above.

achristensen07 commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -123,6 +123,18 @@ struct FontPlatformDataAttributes {
 #endif
 
+#if USE(SKIA)
+    FontPlatformDataAttributes(float size, FontOrientation orientation, FontWidthVariant widthVariant, TextRenderingMode textRenderingMode, bool syntheticBold, bool syntheticOblique, Vector<hb_feature_t>&& features)
```
It might be nice to share the common pieces, but this is probably also fine.  This follows an existing pattern.

donny-dont committed:
Generate serialization of Skia Font implementation
https://bugs.webkit.org/show_bug.cgi?id=273445

Reviewed by Alex Christensen.

Add `WebCoreArgumentCodersSkia.serialization.in` which generates serialization
of Skia's font implementation. Follow along with the CoreText implementation
to reuse as much of their implementation as possible.

* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformDataAttributes::FontPlatformDataAttributes):
* Source/WebCore/platform/graphics/skia/FontCustomPlatformDataSkia.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::FontCustomPlatformData::serializedData const):
* Source/WebCore/platform/graphics/skia/FontPlatformDataSkia.cpp:
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::platformDataInit):
(WebCore::FontPlatformData::fromIPCData):
(WebCore::FontPlatformData::toIPCData const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::shouldUseLockdownFontParser const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreFont.serialization.in:
* Source/WebKit/Shared/harfbuzz/WebCoreArgumentCodersHarfBuzz.serialization.in: Copied from Source\WebKit\Shared\WebCoreFont.serialization.in.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
(IPC::ArgumentCoder<WebCore::Font>::encodePlatformData): Deleted.
(IPC::ArgumentCoder<WebCore::Font>::decodePlatformData): Deleted.
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::encodePlatformData): Deleted.
(IPC::ArgumentCoder<WebCore::FontPlatformData::Attributes>::decodePlatformData): Deleted.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.serialization.in: Copied from Source\WebKit\Shared\WebCoreFont.serialization.in.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):

Canonical link: https://commits.webkit.org/286581@main
